src/Build/BackEnd/Components/RequestBuilder/Lookup.cs

```diff
@@ -658,17 +659,41 @@ internal void AddNewItemsOfItemType(string itemType, ICollection<ProjectItemInst
             IEnumerable<ProjectItemInstance> itemsToAdd = group;
             if (doNotAddDuplicates)
             {
-                // Remove duplicates from the inputs.
-                itemsToAdd = itemsToAdd.Distinct(ProjectItemInstance.EqualityComparer);
 
                 // Ensure we don't also add any that already exist.
                 var existingItems = GetItems(itemType);
+              
+                var existingItemsHashSet = existingItems.ToHashSet(ProjectItemInstance.EqualityComparer);
+
                 if (existingItems.Count > 0)
                 {
-                    itemsToAdd = itemsToAdd.Where(item => !existingItems.Contains(item, ProjectItemInstance.EqualityComparer));
+
+                    var deduplicatedItemsToAdd = new List<ProjectItemInstance>();
+                    foreach (var item in itemsToAdd)
+                    {
+                        if (existingItemsHashSet.Add(item))
+                        {
+                            deduplicatedItemsToAdd.Add(item);
+                        }
+                    }
+                    itemsToAdd = deduplicatedItemsToAdd;
+                }
+                else
+                {
+                    // Remove the duplicates in case we're not concerned with the existing items.
+                    itemsToAdd = itemsToAdd.Distinct(ProjectItemInstance.EqualityComparer);
                 }
             }
 
+            if (doNotAddDuplicates)
+            {
+                logFunction?.Invoke(itemsToAdd.ToList());
```

Comment:
ToList() is not necessary here. We already know exactly that the list is created by us so there's no need to reallocate it.

As I said before, this is an extremely sensitive path allocation-wise, so every allocation needs to be reasoned about. When working with MSBuild, the scale is always bigger than what you think it is. This is going to allocate terabytes and terabytes of memory and burn gigawatts of energy worldwide.

Reply:
```
  if (doNotAddDuplicates)
            {
                logFunction?.Invoke(itemsToAdd.ToList());
```
itemsToAdd is a HashSet due to the optimization we just introduced.
In the other branch, we're using the `asList` check to avoid allocation whenever possible.
is there anything I missed please?

Reply:
Yes, there's some confusion: itemsToAdd is never a HashSet. It is assigned from deduplicatedItemsToAdd (which is a List) on line 678.

In the case where existingItems is empty you're allocating an empty HashSet via a ToHashSet call, and then calling Distinct, which is another hidden allocation of a HashSet. I want to avoid this double allocation.

Let's unify both branches (there's no need to differentiate based on existingItems.Count), this will only allocate a single HashSet for all cases, and we can remove the ToList() in the call to logFunction.

Reply:
Ah, sorry, my bad.
There is still the[ this question](https://github.com/dotnet/msbuild/pull/10820#discussion_r1821312034): 
don't we introduce another duplication(although at a different place) if we remove the if branch due to both building the list and hash set instead of just doing the .Distinct call (which only builds the hash set)?

Reply:
As per my comment to the linked question.